File: backend_disk.go
package bbq

import (
	"bytes"
	"encoding/binary"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"io/fs"
	"math"
	"math/rand"
	"os"
	"path/filepath"

	"github.com/RoaringBitmap/roaring"
	"github.com/barakmich/mmap-go"
)

const defaultVecsPerFile = 200000

type DiskBackend[V any] struct {
	dir          string
	metadata     diskMetadata
	quantization Quantization[V]
	vectorPages  map[int]mmap.MMap
	vectorFiles  map[int]*os.File
	token        uint64
}

type diskMetadata struct {
	Dimensions   int    `json:"dimensions"`
	Quantization string `json:"quantization"`
	VecsPerFile  int    `json:"vecs_per_file"`
	VecFiles     []int  `json:"vec_files"`
}

var _ IndexBackend = &DiskBackend[Vector]{}

func NewDiskBackend[V any](directory string, dimensions int, quantization Quantization[V]) (*DiskBackend[V], error) {
	token := rand.Uint64()
	if token == 0 {
		token = 1
	}
	be := &DiskBackend[V]{
		dir: directory,
		metadata: diskMetadata{
			Dimensions:   dimensions,
			Quantization: quantization.Name(),
			VecsPerFile:  defaultVecsPerFile,
		},
		quantization: quantization,
		token:        token,
		vectorPages:  make(map[int]mmap.MMap),
		vectorFiles:  make(map[int]*os.File),
	}
	err := be.openFiles()
	if err != nil {
		return nil, err
	}
	return be, nil
}

func (d *DiskBackend[V]) Close() error {
	err := d.Sync()
	if err != nil {
		return err
	}
	for _, mm := range d.vectorPages {
		err := mm.Unmap()
		if err != nil {
			return err
		}
	}
	for _, f := range d.vectorFiles {
		err := f.Close()
		if err != nil {
			return err
		}
	}
	return d.saveMetadata()
}

func (d *DiskBackend[V]) Sync() error {
	for _, mm := range d.vectorPages {
		err := mm.FlushAsync()
		if err != nil {
			return err
		}
	}
	return nil
}

func (d *DiskBackend[V]) openFiles() error {
	_, err := os.Stat(d.dir)
	if errors.Is(err, fs.ErrNotExist) {
		return d.createNew()
	} else if err != nil {
		return err
	}

	_, err = os.Stat(filepath.Join(d.dir, "metadata.json"))
	if errors.Is(err, fs.ErrNotExist) {
		return d.createNew()
	} else if err != nil {
		return err
	}

	f, err := os.Open(filepath.Join(d.dir, "metadata.json"))
	if err != nil {
		return err
	}
	defer f.Close()
	err = json.NewDecoder(f).Decode(&d.metadata)
	if err != nil {
		return err
	}

	for _, k := range d.metadata.VecFiles {
		f, err := os.OpenFile(mkPageFilepath(d.dir, k), os.O_RDWR, 0755)
		if err != nil {
			return err
		}
		d.vectorFiles[k] = f
		mm, err := mmap.Map(f, mmap.RDWR, 0)
		if err != nil {
			return err
		}
		d.vectorPages[k] = mm
	}
	return nil
}

func (d *DiskBackend[V]) createNew() error {
	err := os.MkdirAll(d.dir, 0755)
	if err != nil {
		return err
	}
	return d.saveMetadata()
}

func (d *DiskBackend[V]) saveMetadata() error {
	f, err := os.Create(filepath.Join(d.dir, "metadata.json"))
	if err != nil {
		return err
	}
	defer f.Close()
	err = json.NewEncoder(f).Encode(d.metadata)
	if err != nil {
		return err
	}
	return nil
}

func (d *DiskBackend[V]) PutVector(id ID, v Vector) error {
	var err error
	key := int(id) / d.metadata.VecsPerFile
	off := int(id) % d.metadata.VecsPerFile
	page, ok := d.vectorPages[key]
	if !ok {
		page, err = d.createPage(key)
		if err != nil {
			return err
		}
	}
	size := d.quantization.LowerSize(d.metadata.Dimensions)
	l, err := d.quantization.Lower(v)
	if err != nil {
		return err
	}
	slice := page[off*size : (off+1)*size]
	return d.quantization.Marshal(slice, l)
}

func (d *DiskBackend[V]) createPage(key int) (mmap.MMap, error) {
	f, err := os.Create(mkPageFilepath(d.dir, key))
	if err != nil {
		return nil, err
	}
	vecsize := d.quantization.LowerSize(d.metadata.Dimensions)
	err = f.Truncate(int64(vecsize * d.metadata.VecsPerFile))
	if err != nil {
		return nil, err
	}
	d.vectorFiles[key] = f
	mm, err := mmap.Map(f, mmap.RDWR, 0)
	if err != nil {
		return nil, err
	}
	d.vectorPages[key] = mm
	d.metadata.VecFiles = append(d.metadata.VecFiles, key)
	err = d.saveMetadata()
	if err != nil {
		return nil, err
	}
	return mm, nil
}

func (d *DiskBackend[V]) ComputeSimilarity(targetVector Vector, targetID ID) (float32, error) {
	v, err := d.quantization.Lower(targetVector)
	if err != nil {
		return 0, err
	}
	target, err := d.GetVector(targetID)
	if err != nil {
		return 0, err
	}
	return d.quantization.Similarity(target, v), nil
}

func (d *DiskBackend[V]) Info() BackendInfo {
	exists := true
	if _, err := os.Stat(filepath.Join(d.dir, "bases")); errors.Is(err, os.ErrNotExist) {
		exists = false
	}

	return BackendInfo{
		HasIndexData: exists,
		Dimensions:   d.metadata.Dimensions,
		Quantization: d.quantization.Name(),
	}
}

func (d *DiskBackend[V]) Exists(id ID) bool {
	key := int(id) / d.metadata.VecsPerFile
	off := int(id) % d.metadata.VecsPerFile
	page, ok := d.vectorPages[key]
	if !ok {
		return false
	}
	size := d.quantization.LowerSize(d.metadata.Dimensions)
	slice := page[off*size : (off+1)*size]
	for _, x := range slice {
		if x != 0x0 {
			return true
		}
	}
	return false
}

func (d *DiskBackend[V]) GetVector(id ID) (v V, err error) {
	key := int(id) / d.metadata.VecsPerFile
	off := int(id) % d.metadata.VecsPerFile
	page, ok := d.vectorPages[key]
	if !ok {
		err = ErrIDNotFound
		return
	}
	size := d.quantization.LowerSize(d.metadata.Dimensions)
	slice := page[off*size : (off+1)*size]
	return d.quantization.Unmarshal(slice)
}

func (d *DiskBackend[V]) SaveBases(bases []Basis, token uint64) (uint64, error) {
	if token == d.token {
		return d.token, nil
	}
	nbuf := make([]byte, 4)
	buf := bytes.NewBuffer(nil)
	for _, b := range bases {
		for _, v := range b {
			for _, s := range v {
				binary.LittleEndian.PutUint32(nbuf, math.Float32bits(s))
				buf.Write(nbuf)
			}
		}
	}
	f, err := os.Create(filepath.Join(d.dir, "bases"))
	if err != nil {
		return 0, err
	}
	defer f.Close()
	_, err = io.Copy(f, buf)
	return d.token, err
}

func (d *DiskBackend[V]) LoadBases() ([]Basis, error) {
	f, err := os.Open(filepath.Join(d.dir, "bases"))
	if err != nil {
		return nil, err
	}
	var out []Basis
	var basis Basis
	var vec Vector
	buf := make([]byte, 4)
	for {
		_, err = f.Read(buf)
		if errors.Is(err, io.EOF) {
			break
		}
		entry := math.Float32frombits(binary.LittleEndian.Uint32(buf))
		vec = append(vec, entry)
		if len(vec) == d.metadata.Dimensions {
			basis = append(basis, vec)
			vec = nil
			if len(basis) == d.metadata.Dimensions {
				out = append(out, basis)
				basis = nil
			}
		}
	}
	return out, nil
}

func (d *DiskBackend[V]) SaveBitmap(basis int, index int, bitmap *roaring.Bitmap) error {
	path := mkBmapFilepath(d.dir, basis, index)
	f, err := os.Create(path)
	if err != nil {
		return err
	}
	defer f.Close()
	_, err = bitmap.WriteTo(f)
	return err
}

func (d *DiskBackend[V]) LoadBitmap(basis int, index int) (*roaring.Bitmap, error) {
	f, err := os.Open(mkBmapFilepath(d.dir, basis, index))
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return nil, nil
		}
		return nil, err
	}
	bm := roaring.NewBitmap()
	_, err = bm.ReadFrom(f)
	return bm, err
}

func mkPageFilepath(basedir string, key int) string {
	buf := make([]byte, 8)
	binary.BigEndian.PutUint64(buf, uint64(key))
	indexStr := hex.EncodeToString(buf)
	return filepath.Join(basedir, fmt.Sprintf("%s.vec", indexStr))
}

func mkBmapFilepath(basedir string, basis int, index int) string {
	buf := make([]byte, 4)
	binary.BigEndian.PutUint16(buf, uint16(basis))
	basisStr := hex.EncodeToString(buf[:2])
	binary.BigEndian.PutUint32(buf, uint32(index))
	indexStr := hex.EncodeToString(buf[:4])
	return filepath.Join(basedir, fmt.Sprintf("%s-%s.bmap", basisStr, indexStr))
}

File: rust/src/backend_disk.rs
use anyhow::Result;
use std::{collections::HashMap, path::PathBuf};

use serde::{Deserialize, Serialize};

use crate::{
    quantization::Quantization, vector_file::VectorFile, Basis, Bitmap, Vector, VectorBackend, ID,
};

#[derive(Default)]
pub struct DiskBackend<Q: Quantization> {
    dir: PathBuf,
    metadata: DiskMetadata,
    vector_files: HashMap<usize, VectorFile<Q>>,
    token: u64,
}

#[derive(Serialize, Deserialize, Default)]
pub(crate) struct DiskMetadata {
    pub dimensions: usize,
    pub quantization: String,
    pub vecs_per_file: usize,
    pub vec_files: Vec<usize>,
}

const DEFAULT_VECS_PER_FILE: usize = 200_000;

impl<Q: Quantization> DiskBackend<Q> {
    pub fn open(path: PathBuf, dimensions: usize) -> Result<Self> {
        let mut token: u64 = rand::random();
        if token == 0 {
            token = 1;
        }
        let mut s = Self {
            dir: path,
            metadata: DiskMetadata {
                dimensions,
                quantization: Q::name().into(),
                vecs_per_file: DEFAULT_VECS_PER_FILE,
                vec_files: Vec::new(),
            },
            token,
            ..Default::default()
        };
        s.open_files()?;
        Ok(s)
    }

    fn open_files(&mut self) -> Result<()> {
        let metadata_path = self.dir.join("metadata.json");
        if !metadata_path.exists() {
            return self.create_new();
        }
        let metadata_contents = std::fs::read_to_string(&metadata_path)?;
        let metadata: DiskMetadata = serde_json::from_str(&metadata_contents)?;
        self.metadata = metadata;
        for vf in self.metadata.vec_files.iter() {
            let vector_file = VectorFile::<Q>::create_or_open(
                self.make_pagefile_path(vf),
                self.metadata.dimensions,
                self.metadata.vecs_per_file,
            )?;
            self.vector_files.insert(*vf, vector_file);
        }
        Ok(())
    }

    fn create_new(&mut self) -> Result<()> {
        std::fs::create_dir_all(self.dir.clone())?;
        self.save_metadata()
    }

    fn save_metadata(&self) -> Result<()> {
        let metadata_path = self.dir.join("metadata.json");
        Ok(serde_json::to_writer(
            &std::fs::File::create(metadata_path)?,
            &self.metadata,
        )?)
    }

    fn make_pagefile_path(&self, key: &usize) -> PathBuf {
        self.dir.join(format!("{:x}.vec", key))
    }
}

impl<Q: Quantization> VectorBackend for DiskBackend<Q> {
    fn put_vector(&mut self, id: ID, v: &Vector) -> Result<()> {
        todo!()
    }

    fn compute_similarity(&self, target: &Vector, target_id: ID) -> Result<f32> {
        todo!()
    }

    fn info(&self) -> crate::backend::BackendInfo {
        todo!()
    }

    fn iter_vector_ids(&self) -> impl Iterator<Item = ID> {
        0..(self.metadata.vecs_per_file * self.metadata.vec_files.len()) as ID
    }

    fn vector_exists(&self, id: ID) -> bool {
        todo!()
    }

    fn close(self) -> Result<()> {
        todo!()
    }

    fn load_bases(&self) -> Result<Option<Vec<Basis>>> {
        todo!()
    }

    fn load_bitmap<B: Bitmap>(&mut self, basis: usize, index: i32) -> Result<Option<B>> {
        todo!()
    }

    fn save_bases(&mut self, _bases: &[Basis]) -> Result<()> {
        Ok(())
    }

    fn save_bitmap(&mut self, _basis: usize, _index: usize, _bitmap: &impl Bitmap) -> Result<()> {
        Ok(())
    }

    fn sync(&self) -> Result<()> {
        for v in self.vector_files.values() {
            v.flush()?
        }
        self.save_metadata()
    }
}
